File: grimdb/sdk/php/src/Entries/Entry.php
<?php declare(strict_types=1);
namespace Grimlocker\Entries;

/**
 * Generic vault entry holding structured key/value fields.
 *
 * Returned by entry.list / entry.read / entry.search.  For typed access
 * use the specialised entry classes (PasswordEntry, SshKeyEntry, …).
 */
class Entry
{
    /**
     * @param array<string,string> $fields
     */
    public function __construct(
        public readonly string $id,
        public readonly string $title,
        public readonly string $category,
        public readonly array  $fields,
        public readonly int    $createdAt = 0,
        public readonly int    $updatedAt = 0,
    ) {}

    public static function fromArray(array $data): self
    {
        return new self(
            $data['id']         ?? '',
            $data['title']      ?? '',
            $data['category']   ?? '',
            $data['fields']     ?? [],
            $data['created_at'] ?? 0,
            $data['updated_at'] ?? 0,
        );
    }

    public function field(string $key): string
    {
        return $this->fields[$key] ?? '';
    }
}

File: grimdb/sdk/java/src/main/java/com/grimlocker/sdk/Entry.java
package com.grimlocker.sdk;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Collections;
import java.util.Map;

/** A single vault entry returned by the Grimlocker daemon. */
@JsonIgnoreProperties(ignoreUnknown = true)
public class Entry {

    @JsonProperty("id")
    public String id;

    @JsonProperty("category")
    public String category;

    @JsonProperty("title")
    public String title;

    @JsonProperty("fields")
    public Map<String, String> fields = Collections.emptyMap();

    @JsonProperty("created_at")
    public long createdAt;

    @JsonProperty("updated_at")
    public long updatedAt;

    public String field(String key) {
        return fields.getOrDefault(key, "");
    }

    @Override
    public String toString() {
        return "Entry{id='" + id + "', category='" + category + "', title='" + title + "'}";
    }
}
